builder/builder.go

```diff
@@ -0,0 +1,77 @@
+package builder
+
+import (
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/docker/docker/daemon"
+	"github.com/docker/docker/image"
+	"github.com/docker/docker/runconfig"
+)
+
+type ImageID string
+
+type Builder interface {
+	// TODO: make this return a reference and remove ImageID
+	Build(Docker, Context) ImageID
+}
+
+type ImageCache interface {
+	GetCachedImage(ImageID, *runconfig.Config) (ImageID, error)
+	Cache(ImageID, *runconfig.Config) error
+}
+
+type Docker interface {
+	// use digest reference instead of name
+	LookupImage(name string) (*image.Image, error)
+	Pull(name string) (*image.Image, error)
+
+	// move daemon.Container to its own package
+	Container(id string) (*daemon.Container, error)
+	Create(*runconfig.Config, *runconfig.HostConfig) (*daemon.Container, error)
+	Remove(id string, cfg *daemon.ContainerRmConfig) error
+	Commit(*daemon.Container, *daemon.ContainerCommitConfig) (*image.Image, error)
+	Copy(c *daemon.Container, destPath string, srcInfo ContextEntry, decompress bool) error
+}
+
+type Context interface {
+	io.Closer
+	Get(path string) (ContextEntry, error)
+	Open(path string) (io.ReadCloser, error)
+	Walk(filepath.WalkFunc) error
+}
+
+type ModifiableContext interface {
+	Context
+	Create(path string) (io.WriteCloser, error)
+	Remove(path string) error
+}
+
+type ContextEntry interface {
+	os.FileInfo
+	Path() string
+	Open() (io.ReadCloser, error)
```

Comment:
@tiborvass I would recommend trying to remove this interface and work directly with `os.FileInfo` and see if you can get it working. My "jury" is still out on whether an `os.FileInfo` should be trusted for more than just its basename but that is the model the stdlib follows and is common to unix systems. This likely comes from the idea that the path is actually contextual, rather than a property of the file system entry itself. For example, the path of an `os.FileInfo` may be completely differently from one context to the next but could have the same `Name()` and inode. Albeit, this consideration may not be valid for this use case, but may help to clean up the data model.

Let me know what your experience here is, as I have a similar problem in continuity.